File: 백준/Gold/17085. 십자가 2개 놓기/십자가 2개 놓기.java
import java.io.*;
import java.util.*;

public class Main {

    static int N, M, answer = 1;
    static char[][] arr;
    static boolean[][] check;

	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new char[N][M];
        check = new boolean[N][M];

        for(int i = 0; i < N; i ++){
            String str = br.readLine();
            for(int j = 0; j < M; j ++){
                char c = str.charAt(j);
                arr[i][j] = c;
            }
        }

        dfs(0, 1, 0);
        System.out.println(answer);
    }

    public static void dfs(int depth, int cur, int cnt){
        if(cnt >= 2 || depth == (N * M) - 1){
            answer = Math.max(answer, cur);
            return;
        }

        int r = depth / M;
        int c = depth % M;

        for(int i = 1; i <= 15; i ++){
            if(installCross(r, c, i)){
                check(r, c, i, true);
                dfs(depth + 1, cur * ((i * 4) + 1), cnt + 1);
                check(r, c, i, false);
            }else break;
        }

        dfs(depth + 1, cur, cnt);
    }

    public static void check(int r, int c, int size, boolean val){
        for(int i = 0; i <= size; i ++){
            check[r + i][c] = val;
            check[r - i][c] = val;
            check[r][c + i] = val;
            check[r][c - i] = val;
        }
    }

    public static boolean installCross(int r, int c, int size){
        for(int i = 0; i <= size; i ++){
            if(r + i >= N || arr[r + i][c] == '.' || check[r + i][c]) return false;
            if(r - i < 0 || arr[r - i][c] == '.' || check[r - i][c]) return false;
            if(c + i >= M || arr[r][c + i] == '.' || check[r][c + i]) return false;
            if(c - i < 0 || arr[r][c - i] == '.' || check[r][c - i]) return false;
        }
        return true;
    }
}

File: 백준/Gold/17085. 십자가 2개 놓기/십자가 2개 놓기.py

answer = 1


def check(depth, size, flag, v):
    r = int(depth / M)
    c = int(depth % M)

    for i in range(0, size + 1):
        v[r+i][c] = flag
        v[r-i][c] = flag
        v[r][c+i] = flag
        v[r][c-i] = flag

def isPossible(depth, size, grid, v):
    r = int(depth / M)
    c = int(depth % M)

    if r - size < 0 or c - size < 0 or r + size >= N or c + size >= M:
        return False
    
    for i in range(0, size + 1):
        if grid[r + i][c] != '#' or grid[r - i][c] != '#' or grid[r][c + i] != '#' or grid[r][c - i] != '#':
            return False
        
        if v[r + i][c] or v[r - i][c] or v[r][c + i] or v[r][c - i]:
            return False

    return True
            
def dfs(depth, grid, v, cur, cnt):
    global answer
    if cnt >= 2 or depth == N * M - 1:
        answer = max(answer, cur)
        return    
    
    size = 1
    while(isPossible(depth, size, grid, v)):
        check(depth, size, True, v)
        dfs(depth + 1, grid, v, cur * (size * 4 + 1), cnt + 1)
        check(depth, size, False, v)
        size += 1

    dfs(depth + 1, grid, v, cur, cnt)

N, M = map(int, input().split())

grid = [[] for _ in range(N)]
v = [[False for _ in range(M)] for _ in range(N)]

for r in range(N):
    grid[r] = input()

dfs(0, grid, v, 1, 0)
print(answer)
